cabal-install/Distribution/Client/CmdRun.hs

```diff
@@ -128,12 +160,209 @@ runAction (configFlags, configExFlags, installFlags, haddockFlags)
+                  . catMaybes
+                  . fmap (foldPlanPackage
+                           (const Nothing)
+                           (\x -> if match x
+                                  then Just x
+                                  else Nothing))
+                  . toList
+  where
+    -- We need to support ghc 7.6, so we don't have
+    -- a sequenceA that works on tuples yet.
+    -- Once we drop support for pre-ftp ghc
+    -- it's safe to remove this.
+    sequenceA' (a, Just b) = Just (a, b)
+    sequenceA' _ = Nothing
+    match :: ElaboratedConfiguredPackage -> Bool
+    match p = matchPackage pkgId p && matchComponent component p
+    matchingExecutable p = atMostOne
+                         $ filter (\x -> Just x == componentString
+                                      || isNothing componentString)
+                         $ executablesOfPackage p
+    componentString = componentNameString =<< component
+    atMostOne [x] = Just x
+    atMostOne _ = Nothing
+    equalPackageIdAndExe (p,c) (p',c') = c==c' && ((==) `on` elabPkgSourceId) p p'
+
+matchPackage :: Maybe PackageId
+             -> ElaboratedConfiguredPackage
+             -> Bool
+matchPackage pkgId pkg =
+  pkgId == Just (elabPkgSourceId pkg)
+  || isNothing pkgId --if the package is unspecified (Nothing), all packages match
+
+matchComponent :: Maybe ComponentName
+               -> ElaboratedConfiguredPackage
+               -> Bool
+matchComponent component pkg =
+  componentString `elem` (Just <$> executablesOfPackage pkg)
+  || isNothing componentString --if the component is unspecified (Nothing), all components match
+  where componentString = componentNameString =<< component
+
+executablesOfPackage :: ElaboratedConfiguredPackage
+                     -> [UnqualComponentName]
+executablesOfPackage p =
+  case exeFromComponent
+  of Just exe -> [exe]
+     Nothing -> exesFromPackage
+  where
+    exeFromComponent =
+      case elabPkgOrComp p
+      of ElabComponent comp -> case compComponentName comp
+                               of Just (CExeName exe) -> Just exe
+                                  _                   -> Nothing
+         _ -> Nothing
+    exesFromPackage = fmap exeName $ executables $ elabPkgDescription p
+
+-- Path construction
+------
+
+-- | The path to the @build@ directory for an inplace build.
+inplaceBinRoot
+  :: DistDirLayout
+  -> ElaboratedSharedConfig
+  -> ElaboratedConfiguredPackage
+  -> FilePath
+inplaceBinRoot layout config package
+  =  distBuildDirectory layout (elabDistDirParams config package)
+ </> "build"
+
+-- | The path to the directory that contains a specific executable.
+binDirectoryFor
+  :: DistDirLayout
+  -> ElaboratedSharedConfig
+  -> ElaboratedConfiguredPackage
+  -> FilePath
+  -> FilePath
+binDirectoryFor layout config package exe = case elabBuildStyle package of
+  BuildAndInstall -> installedBinDirectory package
+  BuildInplaceOnly -> inplaceBinRoot layout config package </> exe
```

Comment:
This logic is duplicated with `inplace_bin_dir` in `Distribution/Client/ProjectPlanning.hs`. It would be good to have this in only one place.